addons/onboarding/tests/test_onboarding.py

```diff
@@ -0,0 +1,214 @@
+        self.assert_onboarding_is_not_done(self.onboarding_1)
+
+    def test_onboarding_completion_per_company(self):
+        """Checks the behavior of onboarding and step states in multi-company setting:
+        the onboarding state has to track the completion of each of its steps, global and
+        per-company, to determine if whether it is completed.
+        """
+        # Completing onboarding as company_1
+        self.assertEqual(self.env.company, self.company_1)
+
+        # Updating onboarding to per-company
+        self.onboarding_1.is_per_company = True
+        # Required after progress reset
+        self.onboarding_1.search_or_create_progress()
+
+        self.onboarding_1_step_1.set_step_just_done()
+        self.assert_step_is_done(self.onboarding_1_step_1)
+
+        self.onboarding_1_step_2.set_step_just_done()
+        self.assert_onboarding_is_done(self.onboarding_1)
+
+        # Completing onboarding as existing company_2
+        self.activate_company(self.company_2)
+        # First access from company_2
+        self.onboarding_1.search_or_create_progress()
+
+        # Blank state for company 2
+        self.assert_step_is_not_done(self.onboarding_1_step_1)
+        self.assert_onboarding_is_not_done(self.onboarding_1)
+
+        # But no change for company 1
+        self.assert_step_is_done(self.onboarding_1_step_1.with_company(self.company_1))
+        self.assert_onboarding_is_done(self.onboarding_1.with_company(self.company_1))
+
+        self.onboarding_1_step_1.set_step_just_done()
+        self.assert_step_is_done(self.onboarding_1_step_1)
+        self.assert_onboarding_is_not_done(self.onboarding_1)
+        self.onboarding_1_step_2.with_company(self.company_2).set_step_just_done()
+        self.assert_step_is_done(self.onboarding_1_step_2)
+        self.assert_onboarding_is_done(self.onboarding_1)
+
+        # create company_3
+        company_3 = self.env.company.create({
+            'currency_id': self.env.ref('base.EUR').id,
+            'name': 'Another Test Company',
+        })
+        self.activate_company(company_3)
+
+        self.onboarding_1.search_or_create_progress()
+
+        self.assert_onboarding_is_not_done(self.onboarding_1)
+
+        # Completing the steps.
+        self.onboarding_1_step_1.set_step_just_done()
+        self.assert_step_is_done(self.onboarding_1_step_1)
+        self.assert_onboarding_is_not_done(self.onboarding_1)
+        self.onboarding_1_step_2.set_step_just_done()
+        self.assert_onboarding_is_done(self.onboarding_1)
+
+        # is_onboarding_closed status is also company-independent
+        self.onboarding_1.close()
+        self.assertTrue(self.onboarding_1.current_progress_id.is_onboarding_closed)
+        self.assertFalse(self.onboarding_1.with_company(self.company_2).current_progress_id.is_onboarding_closed)
+
+    def assert_step_is_done(self, step, also_with_company=None):
```

Comment:
We could probably move the assert in a small common test class, that way it is easily extensible or re-usable in sub modules :) .